user/tests/unit/pool_allocator.cpp

```diff
@@ -0,0 +1,261 @@
+
+            while(!blocks.empty()) {
+                pool.deallocate(blocks.back());
+                blocks.pop_back();
+            }
+
+            CHECK(pool.available() == size);
+            CHECK(pool.freeList() == nullptr);
+            CHECK(pool.freeListEnd() == nullptr);
+        }
+
+        SECTION("Random order deallocations") {
+            std::vector<void*> blocks;
+
+            bool ok = true;
+            while (ok) {
+                void* p = pool.allocate(1);
+                if (p == nullptr) {
+                    ok = false;
+                    break;
+                }
+
+                blocks.push_back(p);
+            }
+
+            CHECK(blocks.size() > 0);
+
+            {
+                std::random_device r;
+                std::default_random_engine e1(r());
+
+                while(!blocks.empty()) {
+                    std::uniform_int_distribution<int> dist(0, blocks.size() - 1);
+                    const int idx = dist(e1);
+                    pool.deallocate(blocks[idx]);
+                    blocks.erase(blocks.begin() + idx);
+                }
+            }
+
+            CHECK(pool.available() == size);
+            CHECK(pool.freeList() == nullptr);
+            CHECK(pool.freeListEnd() == nullptr);
+        }
+
+        SECTION("Allocations from free list") {
+            std::vector<void*> blocks;
+
+            bool ok = true;
+            while (ok) {
+                void* p = pool.allocate(1);
+                if (p == nullptr) {
+                    ok = false;
+                    break;
+                }
+
+                blocks.push_back(p);
+            }
+
+            CHECK(blocks.size() > 0);
+            CHECK(pool.available() < (sizeof(uintptr_t) * 3));
+
+            const size_t isize = blocks.size();
+            for (unsigned i = 0; i < isize; i++) {
+                void* po = blocks[0];
+                pool.deallocate(po);
+                blocks.erase(blocks.begin());
+                CHECK(pool.allocate(3 * sizeof(uintptr_t)) == nullptr);
+                void* p = pool.allocate(1);
+                CHECK(p != nullptr);
+                CHECK(p == po);
+                blocks.push_back(p);
+            }
+
+            {
+                std::random_device r;
+                std::default_random_engine e1(r());
+
+                while(!blocks.empty()) {
+                    std::uniform_int_distribution<int> dist(0, blocks.size() - 1);
+                    const int idx = dist(e1);
+                    pool.deallocate(blocks[idx]);
+                    blocks.erase(blocks.begin() + idx);
+                }
+            }
+
+            CHECK(pool.available() == size);
+            CHECK(pool.freeList() == nullptr);
+            CHECK(pool.freeListEnd() == nullptr);
+        }
+    }
+
+}
+
+TEST_CASE("SimpleAllocedPool") {
+    Mocks mocks;
+
+    testPool<TestSimpleAllocedPool>(nullptr, DEFAULT_POOL_SIZE);
+}
+
+TEST_CASE("SimpleStaticPool") {
```

Comment:
Since the pool is part of test code it's not strictly necessary to test it, although the benefit is that it rules out where failures in the tests come from (pointing to errors in the code under test rather than the test mocks.)